### PR TITLE
Fixed StringWidth() implementation by using proper Unicode grapheme cluster segmentation. Fixes #28

### LINKED ISSUE #28: Regional Indicators (Flags) and Grapheme Clusters
Here's a short example that illustrates an issue with flags (or "regional indicators"):

```go
fmt.Println(runewidth.StringWidth("🇩🇪")) // Should be "2", outputs "4".
```

The flag consists of two code points which are processed separately by `runewidth`. But most modern systems will combine them into one flag emoji.

This is part of a larger topic which I describe in more detail here: gdamore/tcell#264. It doesn't just affect flags but also characters in e.g. Arabic and Korean where there are more sophisticated rules than "combining characters" and zero-width joiners (which you added with #20).

I don't know exactly how you calculate the widths of characters. I'm also not sure how you would solve flags as well as some of the other rules described in the [Unicode specification](http://unicode.org/reports/tr29/#Grapheme_Cluster_Boundary_Rules) but it would sure be nice as printing these flags currently gives me trouble in [`tview`](https://github.com/rivo/tview). There have been multiple issues asking for better support for different languages and emojis so it seems that there are quite a few people who use the terminal with these characters.

(Maybe my new package [`uniseg`](https://github.com/rivo/uniseg) can help you here.)

### PR DESCRIPTION
**Pull request by rivo**
This introduces an implementation of StringWidth() using Unicode grapheme clusters which should be the correct way to split a string into its individual characters. The built-in assumption is that if we have combined runes (emojis, flags etc.), their width is the width of the first non-zero-width rune. Many of these combined runes were previously not handled correctly by this package.

Please note:

- This introduces a dependency: https://github.com/rivo/uniseg/ (I don't see another way to make this work short of copying most code from `rivo/uniseg` over.)
- No special zero-width joiner (ZWJ) handling is required anymore. Because code and especially configuration options related to ZWJ were thus removed, this branch is not backwards-compatible to the previous commit.
- The changes result in a failure of the `TestStringWidth` test case but only the part where `EastAsianWidth = true`. I'm not very familiar with this flag so I don't know how to fix that. You may want to review this.

### CODE REVIEW

**Comment by mattn** [comment]
Thanks your contribution. I'm thinking there are terminals which does not support ZWJ yet. So I prefer to disable ZWJ.

**Comment by mattn** [comment]
How to add enable/disable ZWJ?

**Comment by rivo** [comment]
Using `rivo/uniseg` means we're doing proper grouping of runes now. There are many more rules than just the handling of a zero-width joiner, see [here](http://unicode.org/reports/tr29/#Grapheme_Cluster_Boundary_Rules). As an example, flags such as 🇯🇵 consist of two runes but don't contain zero-width joiners, they follow other rules.

If you want to support terminals that don't render these characters, the question is _how_ do they render them? If the difference is that some terminals render characters rune-by-rune, then a flag like `UnicodeSegmentation` would be more appropriate.

But singling out the zero-width joiner rule (which is only one of 13 rules) is probably not very useful.

Let me know what you think about this.

**Comment by mattn** [comment]
Most of terminals which does not support ZWJ displays Unicode characters as what we can see. For example, `👪` is displayed as  `👨‍👩‍👦`.

**Comment by rivo** [comment]
So you're saying that these terminals support all kinds of Unicode characters (e.g. flags, Hangul, spacing/prepend marks) but ZWJ is the one they don't support?

**Comment by mattn** [comment]
For example, gnome-terminal displays 👪 as 👨‍👩‍👦.

**Comment by mattn** [comment]
Also ZeroWidthJoiner is alread used in some projects.

https://github.com/4avka/tview/blob/d7a9d6c70ab205d578fc64f02f4e681ccb5b5180/util.go#L56

**Comment by rivo** [comment]
What about `뢴`, `🇩🇪`, `ö`, or `🏳️‍🌈`?

(`tview` is my project. I will remove the ZWJ handling. I only used it because`go-runewidth` required setting it.)

**Comment by mattn** [comment]
I'll check it in later. But probably

뢴 is 2 cells
🇩🇪 is 2 cells
ö is 1 or 2 cells
🏳️‍🌈 is 4 cells

**Comment by rivo** [comment]
Ok. Let me know when you got it.

> For example, gnome-terminal displays 👪 as 👨‍👩‍👦.

Actually, here on GitHub, your answer suggests gnome terminal uses ZWJ properly. Both emojis look identical here.

**Comment by mattn** [comment]
What I look is:

image

**Comment by rivo** [comment]
Interesting. Well, let me know what the result of these other characters is.

If we need to be able to turn off individual Unicode breaking rules, then that functionality needs to be implemented in `uniseg` as well. Actually, in my opinion, it would even be better to set these flags not here in `go-runewidth` but in `uniseg` because these are grapheme breaking flags and that's the package that does it. It would then automatically affect your package, too.

I'll wait for the widths of the characters you wanted to look up and we'll see then how to proceed.

**Comment by alecrabbit** [comment]
> What about 뢴, 🇩🇪, ö, or 🏳️‍🌈? How do these display in a gnome terminal?

If it helps here how these chars look in my terminal(Ubuntu 18.04/Gnome)

image

**Comment by mattn** [comment]
@rivo Can you revert removing ZeroWidthJoiner?

**Comment by mattn** [comment]
Any thought?

**Comment by rivo** [comment]
Apologies for the late reply. And thanks to @alecrabbit for providing the screenshot. Well, reintroducing the `ZeroWidthJoiner` (ZWJ)  flag will not solve the issue with the family icon  (👪) because it doesn't even contain a ZWJ code point (your own package, prior to this PR, gives 2 as a width, no matter how the flag is set). As mentioned before, the issue is larger than just the zero-width joiner topic. I.e. reverting will not solve much.

Also, the `rivo/uniseg` package has no way to disable individual rules so I don't even know how I would implement a ZWJ deactivation. Especially because it could have side effects on other rules.

If you insist on keeping that flag in there, I think you'll need to ignore this pull request.

I think it's generally difficult or even impossible to make this package work on all platforms. One would have to keep a database of how the different platforms render different characters. Since new emojis are introduced all the time, that's whole bunch of work. And it seems that not even popular packages get this right.

Here's iTerm on macOS 10.14:

image

Same with Apple's Terminal:

image

In short, I have no solution. There will always be someone who will complain. Unless someone is willing to put the effort into building a huge database of terminals+OSs and how they render characters, there will be mistakes.

But personally, I think this PR gets us closer to the correct way to handle this than the current version.

**Comment by mattn** [comment]
Could you please fix conflict?

**Comment by rivo** [comment]
Conflicts are resolved. However, some tests still fail. Specifically "TestStringWidth" (see my last bullet point in my opening comment above) and "TestZeroWidthJoiner". Regarding "TestZeroWidthJoiner", the problem appears *not* to be the zero-width joiner but the fact that the white flag 🏳️ (U+1F3F3) is assigned a width of 1 although its actual width is 2.

My terminal (iTerm2 on macOS) seems to make the same mistake:

image

And so is Apple's native Terminal application:

image

So this seems to be a problem across applications. In `go-runewidth`, U+1F3F3 is not contained in the "doublewidth" table:

https://github.com/mattn/go-runewidth/blob/588506649b41f3e5b80a725a9bea447106b4f0d3/runewidth_table.go#L50

I believe it should be in there. If you change the end of this line to

```go
{0x1F3F3, 0x1F3F4}
```

the test will pass. (I can do it in this PR if you want.)

**Comment by mattn** [comment]
>the test will pass. (I can do it in this PR if you want.)

Yes, please

**Comment by rivo** [comment]
>> the test will pass. (I can do it in this PR if you want.)

> Yes, please

I made that change and  also fixed the test case failures resulting from it.

I also fixed the `Truncate()` function which had problems (see [here](https://github.com/rivo/tview/issues/429#issuecomment-625393756) for details).

Please note that as mentioned way above, the `TestStringWidth` test case still fails:

```
=== RUN   TestStringWidth
    TestStringWidth: runewidth_test.go:246: StringWidth("■㈱の世界①") = 10, want 12 (EA)
    TestStringWidth: runewidth_test.go:246: StringWidth("スター☆") = 7, want 8 (EA)
    TestStringWidth: runewidth_test.go:246: StringWidth("つのだ☆HIRO") = 11, want 12 (EA)
```

I'm not  familiar with the `EastAsianWidth` flag so I hope you'll be  able to fix this.  If you need my input on  this, please let me know.

**Comment by mattn** [comment]
Thanks. Could you please fix conflicts?

**Comment by rivo** [comment]
Could you please tell me what changes were made in the main branch? I cannot fix the conflicts if I don't understand what changes were made in parallel.

**Comment by rivo** [comment]
Thanks. This still leaves the failing test regarding the `EastAsianWidth` flag. Please see my message above.

**Comment by rivo** [comment]
Your merge conflict resolutions removed the necessary changes that I made, which is why your test failed. I made them again. Tests pass now.

**Comment by mattn** [comment]
Test is still failing.

https://travis-ci.org/github/mattn/go-runewidth/jobs/687468815

**Comment by rivo** [comment]
I don't know why this is. I fixed these tests. It works for me, see here:

image

**Comment by nojus297** [comment]
`Travis-ci` is running `go generate`, which results in regeneration of `runewidth_table.go`. This changes back one value on line 52 that was changed in this PR and the test fails.

**Comment by dolmen** [comment]
With this change this package isn't about `rune` (as defined in the [Go spec](https://golang.org/ref/spec#Rune_literals)) anymore but about graphemes. A package name change would be appropriate.

**Comment by rivo** [comment]
Thanks @nojus297 for this hint. It took me quite a long time to figure it out. It's a surprisingly difficult issue and it looks like there is not satisfying solution at the moment. So here we go:

`go-runewidth` uses the [`doublewidth` table](https://github.com/mattn/go-runewidth/blob/14e809f6d78fcf9f48ff9b70981472b64c05f754/runewidth_table.go#L23) to check if a rune has a width of 2. That table is generated by extracting all runes from the [East Asian Width table](https://unicode.org/Public/13.0.0/ucd/EastAsianWidth.txt) that have a property "W" which means "wide".

The white flag 🏳️ (which is the basis for the rainbow flag 🏳️‍🌈) is not contained in that table. It's probably new. The way I understand it, the East Asian Width table only includes emojis for historical reasons (East Asian code points were used when emojis were introduced).

[Unicode Annex #11 "East Asian Width"](http://www.unicode.org/reports/tr11/tr11-38.html) says (Section 5, "Recommendations"):

> [UTS51] emoji presentation sequences behave as though they were East Asian Wide, regardless of their assigned East_Asian_Width property value.

I believe `go-runewidth` should actually check the "emoji presentation table" and return a width of 2 if a rune is contained. However, there are two problems with this:

1. `go-runewidth` does not maintain such a table. It only uses the Unicode "Extended_Pictographic" table but that table also contains emojis that have a width of 1.
1. The white flag is not even contained in the Unicode "Emoji_Presentation" table. I don't know if this is a bug or if the recommendation I quoted above is expected to fail for some cases.

In any case, for now, I have changed the test such that for the rainbow flag, a width of 1 is expected. This appears to be in line with the official Unicode information I could find. And as mentioned further above, that's how the terminals also treat it. It's not ideal because obviously, the actual flag is wider than one character. But a solution would probably require the Unicode consortium to make changes (or at least clarify).

All checks are passing now.

**Comment by rivo** [comment]
Any thoughts?

**Comment by mattn** [comment]
As I mentioned above, I don't want to remove ZeroWidthJoiner. So any part using uniseg should be separeted with flag ZeroWidthJoiner.

**Comment by rivo** [comment]
As I mentioned, my project `tview` is the only one using `ZeroWidthJoiner`. (I searched for it.) That's because I was the one who requested it way back. No one else is using it.

As soon as this PR is merged, I will remove `ZeroWidthJoiner` from `tview` and then nobody will use it anymore.

**Comment by mattn** [comment]
>As I mentioned, my project tview is the only one using ZeroWidthJoiner. (I searched for it.) That's because I was the one who requested it way back. No one else is using it.

Ah, I see. Thanks your explaining. Looks good to me. I'll check this soon.

**Comment by mattn** [comment]
Thank you!

**Comment by rivo** [comment]
Thank you!
